### PR TITLE
1.1 ref sim ci workflow

### PR DESCRIPTION
**Pull request by burdorfmitchell**
Implemented a new ci workflow which regression tests and benchmarks the less intensive reference simulations using pytest and pytest-benchmark. The historical simulation output is stored on the Brown Digital Repository. Benchmarking data is stored on the gh-pages branch.

## Description

Added the necessary files to run many of the first generation reference simulations to src/pyuvsim/data:

src/pyuvsim/data/baseline_lite.csv
src/pyuvsim/data/bl_lite_gauss.yaml
src/pyuvsim/data/bl_lite_uniform.yaml
src/pyuvsim/data/mwa88_nocore_config_MWA.yaml
src/pyuvsim/data/mwa88_nocore_config_gauss.yaml
src/pyuvsim/data/test_catalogs/letter_R_12pt_2458098.38824015.txt
src/pyuvsim/data/test_catalogs/mock_catalog_heratext_2458098.38824015.txt
src/pyuvsim/data/test_catalogs/two_distant_points_2458098.38824015.txt
src/pyuvsim/data/test_config/obsparam_ref_1.1_gauss.yaml
src/pyuvsim/data/test_config/obsparam_ref_1.1_mwa.yaml
src/pyuvsim/data/test_config/obsparam_ref_1.1_uniform.yaml
src/pyuvsim/data/test_config/obsparam_ref_1.2_gauss.yaml
src/pyuvsim/data/test_config/obsparam_ref_1.2_uniform.yaml
src/pyuvsim/data/test_config/obsparam_ref_1.3_gauss.yaml
src/pyuvsim/data/test_config/obsparam_ref_1.3_uniform.yaml

These files are then used to run the reference simulations via the test_run_sim pytest method in the added tests/test_run_ref.py file, which has been configured to parametrize command line input to determine which reference simulation to run. The pytest method additionally benchmarks the reference simulation, and downloads older reference simulation output data from the Brown Digital Repository for regression testing.

The pytest method is called from the command line via a new ci workflow currently named "Run Compare Post Benchmark", which runs each reference simulation on a single core in parallel (depending on the amount of github runners available). The benchmark output for each reference simulation is stored as an artifact. Another job in the workflow is then ran which loads all the stored artifacts from the current workflow, concatenates the benchmarking data to a single file, then runs github-action-benchmark.

The github-action-benchmark action loads previous benchmark data from gh-pages, compares it with the current benchmark output, and provides an alert if a performance regression has occurred. The action then pushes the latest benchmark data to gh-pages ONLY if it is a push to main.

Two new dependencies have been added to the environment.yaml and mentioned in the README -- pytest-benchmark and requests. The docs have not been updated, but the README attempts to clarify that serious benchmarking should be done via the benchmarking directory and that this is a single core regression test.

Currently the ci workflow runs on push and pull request as the workflow_run webhook event did not seem to be triggering appropriately, but after merging with main it is likely this can be swapped to running after a successful completion of the Tests workflow. There seems to be roughly enough available runners to manage either way.

## Motivation and Context

The reference simulations have had changes in output over time -- to the extent that some simulations stopped passing an equality check with their own older output. As the output change occurred over many years with minimal oversight, a significant amount of effort was required to confirm the physical consistency of pyuvsim simulations. By incorporating basic regression testing of the reference simulations into the ci workflow, we can identify any time a change in simulation output occurs, even if the cause of the change is external to pyuvsim. We can additionally maintain code performance through benchmarking.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Reference simulation update or replacement
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [x] Build or continuous integration change

## Checklist:

For all pull requests:
- [x] I have read the [contribution guide](CONTRIBUTING.md).
- [x] My code follows the code style of this project.

Documentation change checklist:
- [ ] Any updated docstrings use the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [x] If this is a significant change to the readme or other docs, I have checked that they are rendered properly on ReadTheDocs. (you may need help to get this branch to build on RTD, just ask!)

Build or continuous integration change checklist:
- [x] If required or optional dependencies have changed (including version numbers), I have updated the readme to reflect this.
- [ ] If this is a new CI setup, I have added the associated badge to the readme and to references/make_index.py (if appropriate).

### CODE REVIEW

File: tests/test_run_ref.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    print(f"num_mismatched: {num_mismatched}")
+
+    # perform equality check between historical and current reference
+    # simulation output
+    assert uv_new == uv_ref
+
+    # fail if any element of data_array is different between new and ref
+    # currently commented out but might be worth implementing and updating ref_sim file each time
+    # the assertion fails for the relevant files
+    # assert (ref_arr == new_arr).all()
```
Comment (bhazelton) [inline tests/test_run_ref.py:215]: I don't think we want an exact equality check for floating point numbers, but I'd be comfortable with pretty tight constraints. The defaults in `np.testing.assert_allclose` might be a good choice, but you might also have a good feel for what that should be.

**Review by bhazelton — COMMENTED** [review]
This is looking great, thank you @burdorfmitchell!

File: tests/test_run_ref.py (hunk 1 above)
Comment (burdorfmitchell) [reply tests/test_run_ref.py]: I have updated the code to use the defaults listed in `np.testing.assert_allclose`: atol=0, rtol=1e-7. I have also removed the commented out exact equality check, and removed the discussion about it at the top of the method.

File: .github/workflows/compare-post-benchmark.yaml [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+      # all jobs should run in parallel
+      matrix:
+        include:
+          - id: 1.1_uniform
+          - id: 1.1_gauss
+          - id: 1.1_mwa
+          - id: 1.2_uniform
+          - id: 1.2_gauss
+          - id: 1.3_uniform
+          - id: 1.3_gauss
```
[the remark this answers is not in the document]
Comment (burdorfmitchell) [reply .github/workflows/compare-post-benchmark.yaml:44]: I tried to implement it but it seems to just pass an array in as github.id (and makes a single job matrix), and so the workflow fails. I will hold off on this for now if that is okay, and look to clean up the redundancy in a future pull request for this ci workflow.

[the remark this answers is not in the document]
Comment (burdorfmitchell) [reply .github/workflows/compare-post-benchmark.yaml:44]: actually the previous comment my implementation was wrong (I accidentally kept the include). I will try again.

[the remark this answers is not in the document]
Comment (burdorfmitchell) [reply .github/workflows/compare-post-benchmark.yaml]: Okay I have now implemented it correctly and the code matches your suggestion.

File: README.md [hunk 3]
```diff
@@ -127,6 +127,10 @@ One other package, pytest-xdist, is not required, but can be used to speed up ru
 the test suite by running tests in parallel. To use it call pytest with the
 ```-n auto``` option.
 
+Two additional packages, pytest-benchmark and requests, are required if one needs to locally run
+single core regression testing of the reference simulations. For more realistic benchmarking at
+any level of scale, see [Benchmarking](https://pyuvsim.readthedocs.io/en/latest/developers.html#benchmarking).
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply README.md:132]: I like what you wrote, but I'm re-thinking having it in the readme, it seems a little too technical. Maybe we should move the details about how to run them to the developer docs and just link to that. We could split the benchmarking section into two subsections, one on running the reference sims via pytest-benchmark and the content that is already there into a new subsection? I think it's fine to wait to add more detail until later.

We don't currently have any discussion about what is in the `data` folder. That could go into developer docs too, but I think it's fine to add that later.

**Review by bhazelton — APPROVED** [review]

**Pull request merged**